adminSiteClient/VariablesAnnotationPage.tsx

```diff
@@ -0,0 +1,928 @@
+                        this.context.admin.setErrorMessage({
+                            title: `Failed to fetch variable annotations`,
+                            content:
+                                stringifyUnkownError(err) ??
+                                "unexpected error value in setErrorMessage",
+                            isFatal: false,
+                        })
+                        return [undefined]
+                    })
+                )
+            })
+        )
+        // Turn the rxJs observable stream into a mobx value
+        const mobxValue = fromStream(observable)
+        // Set up the autorun to run an action to update the dependendencies
+        const disposer = autorun(() => {
+            const currentData = mobxValue.current
+
+            if (currentData !== undefined) {
+                runInAction(() => {
+                    console.log("Data fetched", currentData)
+                    this.resetViewStateAfterFetch()
+                    this.currentPagingOffset = this.desiredPagingOffset
+                    this.richDataRows = currentData.variables.map(
+                        parseVariableAnnotationsRow
+                    )
+                    this.numTotalRows = currentData.numTotalRows
+                })
+            }
+        })
+        // Add the disposer to the list of things we need to clean up on unmount
+        this.disposers.push(disposer)
+
+        this.getFieldDefinitions()
+    }
+
+    @action.bound private resetViewStateAfterFetch(): void {
+        this.undoSteps = []
+        this.redoSteps = []
+        this.selectedRow = undefined
+        this.grapherElement = undefined
+    }
+    @action.bound private loadGrapherJson(json: any): void {
+        const newConfig: GrapherProgrammaticInterface = {
+            ...json,
+            bounds: new Bounds(0, 0, 500, 400),
+            getGrapherInstance: (grapher: Grapher) => {
+                this.grapher = grapher
+            },
+        }
+        if (this.grapherElement) {
+            {
+                this.grapher.setAuthoredVersion(newConfig)
+                this.grapher.reset()
+                this.grapher.updateFromObject(newConfig)
+                this.grapher.downloadData()
+            }
+        } else this.grapherElement = <Grapher {...newConfig} />
+    }
+
+    @action private updatePreviewToRow(row: number): void {
+        const { richDataRows } = this
+        console.log("updating preview to row", row)
+        if (richDataRows === undefined) return
+
+        // Get the grapherConfig of the currently selected row and then
+        // merge it with the necessary partial information (e.g. variableId field)
+        // to get a config that actually works in all cases
+        const config = richDataRows[row].grapherConfig
+        const finalConfigLayer = getFinalConfigLayerForVariable(
+            richDataRows[row].id
+        )
+        const mergedConfig = merge(config, finalConfigLayer)
```

Comment:
That is a valid concern - I think in practice the way things are used it shouldn't affect us much and in the mid term I would like to migrate the grapher config structure to a place where it is more "correct by construction" to help with this. E.g. if currently we have 3 arrays that have to be in sync it would be better to have one array with a nested structure of 3 values so that this problem can't come up. Do you think it is better to invest more into verification here that this value is correct?

Reply:
Sounds valid to me, and an absolute thumbs up to `it would be better to have one array with a nested structure of 3 values so that this problem can't come up`!